spec.html

```diff
@@ -38326,6 +38326,126 @@ <h1>`Promise.all` Resolve Element Functions</h1>
         </emu-clause>
       </emu-clause>
 
+      <emu-clause id="sec-promise.allsettled">
+        <h1>Promise.allSettled ( _iterable_ )</h1>
+        <p>The `allSettled` function returns a promise that is fulfilled with an array of promise state snapshots, but only after all the original promises have settled, i.e. become either fulfilled or rejected. It resolves all elements of the passed iterable to promises as it runs this algorithm.</p>
+        <emu-alg>
+          1. Let _C_ be the *this* value.
+          1. If Type(_C_) is not Object, throw a *TypeError* exception.
+          1. Let _promiseCapability_ be ? NewPromiseCapability(_C_).
+          1. Let _iteratorRecord_ be GetIterator(_iterable_).
+          1. IfAbruptRejectPromise(_iteratorRecord_, _promiseCapability_).
+          1. Let _result_ be PerformPromiseAllSettled(_iteratorRecord_, _C_, _promiseCapability_).
+          1. If _result_ is an abrupt completion, then
+            1. If _iteratorRecord_.[[Done]] is *false*, set _result_ to IteratorClose(_iteratorRecord_, _result_).
+            1. IfAbruptRejectPromise(_result_, _promiseCapability_).
+          1. Return Completion(_result_).
```

Comment:
why is `Completion()` needed here? we already know it's not an abrupt completion, so isn't it already a normal one?
```suggestion
          1. Return _result_.
```

Reply:
Per <https://tc39.es/ecma262/#sec-implicit-completion-values>:

> The abstract operation `Completion(completionRecord)` is used to emphasize that a previously computed Completion Record is being returned. The `Completion` abstract operation takes a single argument, `completionRecord`, and performs the following steps: [...]

Reply:
As an aside, I’d love to come up with a ?-like pattern around all these IfAbruptRejectPromise combos someday